elastic/beats
# [Metricbeat][Kubernetes] Update state_* metricsets to be compatible with supported kubernetes versions

**This PR is going to be split. Please see https://github.com/elastic/beats/pull/34432.**

## What does this PR do?
Update the metrics of the `state_*` metricsets.

**WARNING**: Any dashboards that use these metrics were not looked into in this PR; `state_container` and `state_resourcequota` metricsets functions were not modified, unlike the others.

## Why is it important?
Kube state metrics versions tested before are no longer supported by the metricbeat Kubernetes module.

Changes:

**New files**:
- Updated the `data.json` examples
- Deleted the previous files for KSM, and added new ones that support versions 1.23.x-1.26.x of K8s
- Added `state_metricset.go` file with all the common functions between the `state_*` metricsets (expect for resource quota and container, as they don't follow the same logic as the others)
- Added new file `ktest.go` for kubernetes related testing functions

**New test**:
`TestStateMetricsFamily` was added to all testing files of the `state_*` metricsets.
The only purpose of it, is to help us remove deprecated metrics. It simply goes over the kube state metrics files, and checks if the family metric exists in it. It does **not** check for labels, as they may not appear in the files, even though they may exist - e.g. if there is no data for that metric.

**Updated file**:
`metricbeat/module/kubernetes/_meta/test/docs/README.md` was changed to make it more clear for us to know how to test the metrics.

Most of the **changes on the metrics/labels** are:

- State container:
   - `kube_pod_container_resource_limits_*_cores` replaced by `kube_pod_container_resource_limits`
   - `kube_pod_container_resource_requests_*_cores` replaced by `kube_pod_container_resource_requests`
   - `kube_pod_container_status_terminated` removed since it can be easily obtained through `kube_pod_container_status_terminated_reason`
   - `kube_pod_container_status_waiting` removed since it can be easily obtained through `kube_pod_container_status_waiting_reason`
   - `kube_pod_container_status_restarts` removed - it does not have data in any of the KSM versions supported by us
   - `reason` added - from `kube_pod_container_status_*_reason`
- State job:
  - `condition` label added
- State node:
  - `kube_node_status_allocatable_*` replaced by `kube_node_status_allocatable`
  - `kube_node_status_capacity_cpu_cores_*` replaced by `kube_node_status_capacity`
  - `condition` label added
  - `status` label added
  - `resource` label added
  - `unit` label added
- State storage class:
  - `volumeBindingMode` removed
  - `reclaimPolicy` removed

You can find more information about the metrics in the README file of each metricset.


## Checklist



## Related issues

- Relates to https://github.com/elastic/beats/pull/34275


## How to test
Please, see [this file](https://github.com/constanca-m/beats/blob/update-kube-state-metrics-tests/metricbeat/module/kubernetes/_meta/test/docs/README.md).
